### PR TITLE
commands/auth_utils: bugfix on config directory creation

### LINKED ISSUE #434: --config with auth login only works correctly with "config" as filename
## Issue
When a custom configuration file is supplied via `--config` to `auth login`, and reading credentials from the given file doesn't succeed, an attempt is made to create the supplied directory path. This path is constructed by stripping a fixed "config" (without a directory separator) from the user-supplied path.

When the custom configuration file doesn't exist, this results in an unnecessary `mkdir` when the given filename ends with the hardcoded value of "config" (case 1 below), or that plus a "is a directory" failure when it doesn't (case 2).

If the custom configuration file *does* exist, the useless `mkdir` is still triggered when the given filename ends with the hardcoded value of "config" (case 3), or a "not a directory" failure when it doesn't (case 4).

## Cause
[The problematic `mkdir` call is issued here](https://github.com/mattermost/mmctl/blob/ae40c25f62d8f7e20e403266ffce3980aad195cf/commands/auth_utils.go#L152). I have no experience in Go, but instead of the `TrimSuffix()` call nested on that line, I think there should be a Go equivalent of a [`dirname`](https://en.wikipedia.org/wiki/Dirname) instead, using the user-supplied path as parameter, and without referencing the value of `configFileName` (the source of the hardcoded "config").

### Case 1

A useless, empty `mmctl_` directory gets created:

    # rm -rf /tmp/mmctl_test/
    # sudo -u mattermost mkdir /tmp/mmctl_test/
    # /usr/bin/sudo -u mattermost /usr/local/bin/mmctl auth --quiet login https://example.com --name example --access-token 123456890abcdefghijklmnopq --config /tmp/mmctl_test/mmctl_config

      credentials for "example": "Personal Access Token@https://example.com" stored

    # ls -lRa /tmp/mmctl_test/
    /tmp/mmctl_test/:
    yhteensä 24
    drwxrwxr-x  3 mattermost mattermost  4096 tammi  30 15:27 .
    drwxrwxrwt 23 root       root       12288 tammi  30 15:27 ..
    drwx------  2 mattermost mattermost  4096 tammi  30 15:27 mmctl_
    -rw-------  1 mattermost mattermost   244 tammi  30 15:27 mmctl_config

    /tmp/mmctl_test/mmctl_:
    yhteensä 8
    drwx------ 2 mattermost mattermost 4096 tammi  30 15:27 .
    drwxrwxr-x 3 mattermost mattermost 4096 tammi  30 15:27 ..

### Case 2

Saving the credentials fails:

    # rm -rf /tmp/mmctl_test/
    # sudo -u mattermost mkdir /tmp/mmctl_test/
    # /usr/bin/sudo -u mattermost /usr/local/bin/mmctl auth --quiet login https://example.com--name example --access-token 123456890abcdefghijklmnopq --config /tmp/mmctl_test/test
    Error: cannot save the credentials: open /tmp/mmctl_test/test: is a directory
    # ls -lRa /tmp/mmctl_test/
    /tmp/mmctl_test/:
    yhteensä 20
    drwxrwxr-x  3 mattermost mattermost  4096 tammi  30 15:31 .
    drwxrwxrwt 23 root       root       12288 tammi  30 15:31 ..
    drwx------  2 mattermost mattermost  4096 tammi  30 15:31 test

    /tmp/mmctl_test/test:
    yhteensä 8
    drwx------ 2 mattermost mattermost 4096 tammi  30 15:31 .
    drwxrwxr-x 3 mattermost mattermost 4096 tammi  30 15:31 ..

### Case 3

A useless, empty `test_` directory is created:

    # rm -rf /tmp/mmctl_test/
    # sudo -u mattermost mkdir /tmp/mmctl_test/
    # sudo -u mattermost touch /tmp/mmctl_test/test_config
    # /usr/bin/sudo -u mattermost /usr/local/bin/mmctl auth login https://example.com --name example --access-token 123456890abcdefghijklmnopq --config /tmp/mmctl_test/test_config

      credentials for "example": "Personal Access Token@https://example.com" stored

    # ls -lRa /tmp/mmctl_test/
    /tmp/mmctl_test/:
    yhteensä 24
    drwxrwxr-x  3 mattermost mattermost  4096 tammi  30 15:58 .
    drwxrwxrwt 22 root       root       12288 tammi  30 15:58 ..
    drwx------  2 mattermost mattermost  4096 tammi  30 15:58 test_
    -rw-rw-r--  1 mattermost mattermost   244 tammi  30 15:58 test_config

    /tmp/mmctl_test/test_:
    yhteensä 8
    drwx------ 2 mattermost mattermost 4096 tammi  30 15:58 .
    drwxrwxr-x 3 mattermost mattermost 4096 tammi  30 15:58 ..

### Case 4

Saving the credentials again fails:

    # rm -rf /tmp/mmctl_test/
    # sudo -u mattermost mkdir /tmp/mmctl_test/
    # sudo -u mattermost touch /tmp/mmctl_test/test
    # /usr/bin/sudo -u mattermost /usr/local/bin/mmctl auth login https://example.com --name example --access-token 123456890abcdefghijklmnopq --config /tmp/mmctl_test/test
    Error: mkdir /tmp/mmctl_test/test: not a directory
    # ls -lRa /tmp/mmctl_test/
    /tmp/mmctl_test/:
    yhteensä 16
    drwxrwxr-x  2 mattermost mattermost  4096 tammi  30 16:00 .
    drwxrwxrwt 22 root       root       12288 tammi  30 16:00 ..
    -rw-rw-r--  1 mattermost mattermost     0 tammi  30 16:00 test

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Fixes #434

### CODE REVIEW

File: commands/auth_utils_test.go [hunk 1]
```diff
@@ -85,6 +86,29 @@ func TestResolveConfigFilePath(t *testing.T) {
...
+
+		testUser.HomeDir = "path/should/be/ignored"
+		SetUser(testUser)
+		extraDir := "extra"
+
+		expected := filepath.Join(tmp, extraDir, "config.json")
+
+		err := os.Setenv("XDG_CONFIG_HOME", "path/should/be/ignored")
+		require.NoError(t, err)
+		viper.Set("config", expected)
```
Comment (agnivade) [inline commands/auth_utils_test.go:102]: Should `--config` be a path to file or can be a directory as well? If yes, can we add a test where we just supply a custom dir ending with `_config` and ensure that a redundant directory does not get created?

Comment (isacikgoz) [reply commands/auth_utils_test.go:102]: It is a file path, the old flag (`--config-path`) was expecting it to be a directory and filename was being determined by the tool itself. That was a little bit problematic approach therefore we replaced it (with `--config`, in fact we emphasized that flag expects a file in the docs and help text).

Comment (agnivade) [reply commands/auth_utils_test.go:102]: Great :)

**Review by agnivade — APPROVED** [review]
Good to go after a test is added

**Pull request merged**